src/include/miopen/reduce_common.hpp

```diff
@@ -0,0 +1,311 @@
+#ifndef GUARD_MIOPEN_REDUCE_COMMON_HPP
+#define GUARD_MIOPEN_REDUCE_COMMON_HPP
+
+#include <half.hpp>
+#include <limits>
+#include <cmath>
+#include <miopen/miopen.h>
+#include <miopen/float_equal.hpp>
+
+#include "bfloat16.hpp"
+
+namespace reduce {
+
+enum ReductionMethod_t
+{
+    Reduce_DirectThreadWise = 1,
+    Reduce_DirectWarpWise   = 2,
+    Reduce_BlockWise        = 3,
+    Reduce_MultiBlock       = 4
+};
+
+// data type conversion
+template <typename T>
+struct type_convert
+{
+    template <typename X>
+    T operator()(X x) const
+    {
+        return static_cast<T>(x);
+    }
+};
+
+template <>
+template <>
+inline float type_convert<float>::operator()<half_float::half>(half_float::half x) const
+{
+    return half_float::half_cast<float>(x);
+};
+
+template <>
+template <>
+inline half_float::half type_convert<half_float::half>::operator()<float>(float x) const
+{
+    return half_float::half_cast<half_float::half>(x);
+};
+
+template <>
+template <>
+inline float type_convert<float>::operator()<bfloat16>(bfloat16 x) const
+{
+    return float(x);
+};
+
+template <>
+template <>
+inline bfloat16 type_convert<bfloat16>::operator()<float>(float x) const
+{
+    return bfloat16(x);
+};
+
+template <typename compType>
+static inline std::function<void(compType&, compType)> ReduceOpFn(miopenReduceTensorOp_t op_)
+{
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_ADD: return ([&](compType& a_, compType b_) { a_ = a_ + b_; });
+
+    case MIOPEN_REDUCE_TENSOR_MUL: return ([&](compType& a_, compType b_) { a_ = a_ * b_; });
+
+    case MIOPEN_REDUCE_TENSOR_MIN:
+        return ([&](compType& a_, compType b_) {
+            if(a_ > b_)
+                a_ = b_;
+        });
+
+    case MIOPEN_REDUCE_TENSOR_MAX:
+        return ([&](compType& a_, compType b_) {
+            if(a_ < b_)
+                a_ = b_;
+        });
+    }
+
+    return (std::function<void(compType&, compType)>{});
+};
+
+template <typename compType>
+static inline std::function<void(compType&, compType, bool& changed)>
+ReduceOpFn2(miopenReduceTensorOp_t op_)
+{
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_MIN:
+        return ([&](compType& a_, compType b_, bool& changed) {
+            if(a_ > b_)
+            {
+                a_      = b_;
+                changed = true;
+            }
+            else
+                changed = false;
+        });
+
+    case MIOPEN_REDUCE_TENSOR_MAX:
+        return ([&](compType& a_, compType b_, bool& changed) {
+            if(a_ < b_)
+            {
+                a_      = b_;
+                changed = true;
+            }
+            else
+                changed = false;
+        });
+
+    case MIOPEN_REDUCE_TENSOR_ADD:
+    case MIOPEN_REDUCE_TENSOR_MUL: return (std::function<void(compType&, compType, bool&)>{});
+    };
+
+    return (std::function<void(compType&, compType, bool&)>{});
+};
+
+template <typename compType>
+static inline compType ReduceOpZeroVal(miopenReduceTensorOp_t op_)
```

Comment:
This is an initializer function?

Reply:
ReduceOpZeroVal() is to get the zero value for a specific reduce op type.  "reduce_common.hpp" is used by both the driver/reduce_driver.cpp and test/reduce_test.cpp.